src/firmware/main/math/polynomial_2d.c

```diff
@@ -0,0 +1,178 @@
+#include "firmware/main/math/polynomial_2d.h"
+
+#include <assert.h>
+#include <math.h>
+#include <stdbool.h>
+#include <stdio.h>
+
+/**
+ * Calculate the L2 norm of the vector at the given t-value
+ *
+ * @param p The polynomial to get the vector from
+ * @param t The t-value to get the vector at
+ * @return The L2 norm of the polynomial at the given t-value
+ */
+#define GENERATE_2D_POLYNOMIAL_CALCULATE_L2_NORM_DEFINITION(N)                           \
+    float calculateL2NormAtValueOrder##N(Polynomial2dOrder##N##_t p, float t)            \
+    {                                                                                    \
+        const float x_value = shared_polynomial1d_getValueOrder##N(p.x, t);              \
+        const float y_value = shared_polynomial1d_getValueOrder##N(p.y, t);              \
+                                                                                         \
+        return sqrt(pow(x_value, 2.0) + pow(y_value, 2.0));                              \
+    }
+GENERATE_2D_POLYNOMIAL_CALCULATE_L2_NORM_DEFINITION(3);
+GENERATE_2D_POLYNOMIAL_CALCULATE_L2_NORM_DEFINITION(2);
+GENERATE_2D_POLYNOMIAL_CALCULATE_L2_NORM_DEFINITION(1);
+GENERATE_2D_POLYNOMIAL_CALCULATE_L2_NORM_DEFINITION(0);
+
+#define GENERATE_2D_POLYNOMIAL_GET_VALUE_FUNCTION_DEFINITION(N)                          \
+    Vector2d_t shared_polynomial2d_getValueOrder##N(Polynomial2dOrder##N##_t p, float t) \
+    {                                                                                    \
+        Vector2d_t result = {                                                            \
+            .x = shared_polynomial1d_getValueOrder##N(p.x, t),                           \
+            .y = shared_polynomial1d_getValueOrder##N(p.y, t),                           \
+        };                                                                               \
+        return result;                                                                   \
+    }
+GENERATE_2D_POLYNOMIAL_GET_VALUE_FUNCTION_DEFINITION(3)
+GENERATE_2D_POLYNOMIAL_GET_VALUE_FUNCTION_DEFINITION(2)
+GENERATE_2D_POLYNOMIAL_GET_VALUE_FUNCTION_DEFINITION(1)
+GENERATE_2D_POLYNOMIAL_GET_VALUE_FUNCTION_DEFINITION(0)
+
+#define GENERATE_2D_POLYNOMIAL_DIFFERENTIATE_FUNCTION_DEFINITION(FROM_ORDER, TO_ORDER)   \
+    Polynomial2dOrder##TO_ORDER##_t shared_polynomial2d_differentiateOrder##FROM_ORDER(  \
```

Comment:
similar to above comment, can we `static_assert` that `TO_ORDER = FROM_ORDER - 1`?

Reply:
Yep, added